fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java

```diff
@@ -95,7 +96,17 @@ public List<OptExpression> transform(OptExpression queryExpression, OptimizerCon
                 }
             }
         } else {
-            mvCandidateContexts = context.getCandidateMvs();
+            mvCandidateContexts.addAll(context.getCandidateMvs());
+        }
+        mvCandidateContexts.removeIf(x -> !x.prune(context, queryExpression));
+        mvCandidateContexts.sort(new MaterializationContext.RewriteOrdering());
+        int numCandidates = context.getSessionVariable().getCboMaterializedViewRewriteCandidateLimit();
+        if (numCandidates > 0 && mvCandidateContexts.size() > numCandidates) {
```

Comment:
check the limit first and then partial sort to fetch the topN?

Reply:
there're two pruning mechanisms:
- pre-prune: prune by `cbo_materialized_view_rewrite_candidate_limit`, choose the TopN candidates
- post-prune: prune by `cbo_materialized_view_rewrite_output_limit`, choose the first N result

so if user doesn't specify the `candiate_limit` but only the `output_limit`, it would do the full sort and choose first N result, even if no pruning for the candidates